kernel/protocol/attest: Add extended attestation support

IT302 opened the pull request:
Add support for requesting an extended attestation package (report + manifest + certificate bundle) for SVSM_ATTEST_SERVICES and SVSM_ATTEST_SINGLE_SERVICE.

The certificate bundle is optional and may be supplied by the host. If unavailable, only the report and manifest are returned.

This enables consumers to obtain a self-contained evidence set without extra host round trips. Linux requests the extended attestation report by default.

stefano-garzarella commented:
@IT302 thanks for the PR. How can we test this PR and how this is related to #719 ?

IT302 commented:
> @IT302 thanks for the PR. How can we test this PR and how this is related to #719 ?

This commit enables extended attestation by including (when present on the host) the additional certificate material needed to verify the report, as described in Section 7.2 of the “Secure VM Service Module for SEV-SNP Guests” specification. This is unrelated to issue #719 .

Current problem: The existing implementation ignores requests for extended attestation and always performs a standard attestation flow. That is technically incorrect: the request parameters should decide whether the SVSM returns only the basic attestation report and manifest (standard) or the report, manifest plus the host-provided certificate objects (extended).

What this PR changes:

Detects when the caller requests extended attestation.
If certificates are available on the host, they are copied into (and returned through) the certificate buffer provided.
Falls back gracefully: if the certificates are not present, the attestation still succeeds, but only the standard components are returned. The SVSM will log a warning in that case.

How to test with Linux TSM (example): (report is a configfs TSM report directory)

```
report=/sys/kernel/config/tsm/report/report0
mkdir -p  $report 
echo svsm > $report/service_provider 
echo c476f1eb-0123-45a5-9641-b4e7dde5bfe3 > $report/service_guid
dd if=/dev/urandom bs=64 count=1 of=$report/inblob
hexdump -C $report/outblob 
ls -l $report/
```

References
https://github.com/virtee/snpguest/blob/main/README.md#extended-attestation-flowchart

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -257,6 +327,35 @@ fn get_attestation_report(nonce: &[u8]) -> Result<Box<SnpReportResponse>, SvsmRe
+
+    // Zero initialized, so
+    // vmpl=0
+    // flags=1: Use VCEK.
```
Why here we use flags=1 and in `get_attestation_report()` flags=0 ?

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -326,10 +477,36 @@ fn attest_single_service(
+        // from the PSP indicating that the certificate buffer is too small.
+        // The required size is in certs_buffer_size.
+        // Per spec, return required size in rdx register and raise an error.
+        let (resp, certs) = match get_attestation_report_extended(hash.as_slice()) {
```
In the commit message we mention

> Add support for requesting an extended attestation package
    (report + manifest + certificate bundle) for SVSM_ATTEST_SERVICES and
    SVSM_ATTEST_SINGLE_SERVICE.

But we call `get_attestation_report_extended()` only in `attest_single_service()`, so how we support SVSM_ATTEST_SERVICES?

IT302 commented:
Added code to support SVSM_ATTEST_SERVICES. Removed an erroneous comment regarding flags and refactored the implementation by extracting the logic into a separate function.

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -36,6 +39,18 @@ const SVSM_ATTEST_SINGLE_SERVICE: u32 = 1;
+// or "SEV-ES Guest-Hypervisor Communication Block Standardization, Revision 2.04"
+// If future specifications update the maximum certificate buffer size, update the
+// MAX_CERTIFICATE_SIZE constant below to match the new value.
+const MAX_CERTIFICATE_SIZE: usize = 0x4000;
```
@tlendacky @joergroedel should we add this into the spec?

IT302 commented:
> > Added code to support SVSM_ATTEST_SERVICES. Removed an erroneous comment regarding flags and refactored the implementation by extracting the logic into a separate function.
> 
> @IT302 thanks for the update, and sorry for the long delay.
> 
> BTW how can I test separately SVSM_ATTEST_SERVICES and SVSM_ATTEST_SINGLE_SERVICE ?
> 
> IIUC [#798 (comment)](https://github.com/coconut-svsm/svsm/pull/798#issuecomment-3334456211) is for SVSM_ATTEST_SINGLE_SERVICE, right?
To test SVSM_ATTEST_SERVICES, use the above but set service-guid to null

```
report=/sys/kernel/config/tsm/report/report0
mkdir -p  $report 
echo svsm > $report/service_provider 
echo 00000000-0000-0000-0000-000000000000 > $report/service_guid
dd if=/dev/urandom bs=64 count=1 of=$report/inblob
hexdump -C $report/outblob 
hexdump -C  $report/manifestblob
```

stefano-garzarella commented:
@IT302 I'm not really an expert on this topic but I'd like to merge this PR.
Can you provide steps that I need to do to provide certificate bundle from the host and check in the guest they are the same.

Would be nice to have a `test_in_svsm` to test it automatically, but for now a step by step should be okay.

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -307,12 +410,99 @@ fn write_report_and_manifest(
         .map_err(|_| SvsmError::Attestation(AttestError::Manifest))?;
 
-    // Does not support certificate currently, so setting certificate size to 0
-    params.rdx = 0;
```
Sorry, I didn't mean to remove this, but to set here (on in another single place) on the right value.

Now for example, if the guest ask for the standard report, where we write it?

IMO we should avoid to spread rdx write all around a code, but have a single point.

IT302 replied:
I understand what you meant originally, but it turns out we don’t actually need it anymore as the PR adds extended attestation handling.

Background:

Per the SVSM spec, a “standard” attestation means the host does not provide a certificate buffer. In that case, the SVSM should not set the certificate buffer length in RDX.
The original implementation did not distinguish between standard and extended attestations and always wrote RDX = 0 to indicate there is no certificate buffer.

stefano-garzarella reviewed (commented):
@IT302 LGTM and I successfully tested, I have just one small question about error path.

IT302 commented on kernel/src/protocols/attest.rs:
```diff
@@ -307,12 +410,99 @@ fn write_report_and_manifest(
+                // The PSP returned an error code indicating that the buffer size is too small.
+                // The required size is in certs_buffer_size.
+                // Per spec, return required size in rdx and raise an error
+                params.rdx = certs_buffer_size;
```
@stefano-garzarella Hold on for a few minutes, I did not push the latest commit.

stefano-garzarella commented on kernel/src/protocols/attest.rs:
```diff
@@ -307,12 +410,99 @@ fn write_report_and_manifest(
+                // The PSP returned an error code indicating that the buffer size is too small.
+                // The required size is in certs_buffer_size but specified as number of 4 KB pages.
+                // Per SVSM spec, return required size (in bytes) in rdx and raise an error
+                params.rdx = certs_buffer_size * 4096;
```
What about using `PAGE_SHIFT` like in https://github.com/coconut-svsm/svsm/blob/346f4862d73ef40c9c9cf7c05cf2c444dff2be3a/kernel/src/greq/driver.rs#L132-L135
